### DISCUSSION TITLE
CrossHair + PyPy Opportunities.

### DISCUSSION THREAD

**Discussion by pschanely** [pschanely/CrossHair#104] [Ideas]
A continuation of this thread about [PyPy and pure Python implementations](https://github.com/pschanely/CrossHair/issues/3#issuecomment-815394439):

> @pschanely If we run CrossHair with pypy does that remove the need for symbolic versions of stuff pypy implements in pure python like `collections.deque`? Are there other benefits of running CrossHair with pypy?

My suspicion is that there is a lot to learn / leverage from PyPy.

**As for reusing pure python implementations of the standard library:**

This warrants more investigation. I bet there are implementations that would be immediately usable by CrossHair that I just haven't considered. Sometimes, the implementation won't be ideal for CrossHair. [This implementation of deque](https://foss.heptapod.net/pypy/pypy/-/blob/release-5.1.2/lib_pypy/_collections.py), for instance, is focused on performance - running CrossHair on it will create much more work for the system than is necessary. You'll find this a common theme in CrossHair implementations: we use very "slow" algorithms because they make solving easier. A great example here: CrossHair containment checks in sets/dictionaries are done by iterating over the contents! Absurd right? Well, turns out that hashing things quickly makes the solver state unmanageable. So we are happy to go through great lengths to avoid hashes.

**Some additional avenues of exploration:**

* Right now, CrossHair is CPython-specific, and doesn't run on PyPy. The major blocker here is some deep internal hacks we do to while bytecode tracing to hot swap functions before we call them. With some work, this logic could be abstracted in a way that could be called from either platform. Possible benefits of this exploration might be: performance benefits, some kind of uses for the reversibility/"time travel" feature, and more flexibility in intercepting/customizing behavior.
* I'm not fully sure about this, but I think PyPy's compilation process is somehow running Python and observing/remembering machine code; that certainly sounds vaguely similar to how CrossHair works. One might fully re-imagine CrossHair as a first-class participant in PyPy. This is probably a sufficiently different approach that it should be a new project.

**Comment by andrew222651** [comment]
Here's some other work related to implementing (parts of) Python in Python: https://github.com/brettcannon/desugar
